test/e2e/aws_eip_for_nlb_test.go

```diff
@@ -0,0 +1,492 @@
+//go:build e2e
+// +build e2e
+
+package e2e
+
+import (
+	"context"
+	"fmt"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	configv1 "github.com/openshift/api/config/v1"
+	"github.com/openshift/api/features"
+	operatorv1 "github.com/openshift/api/operator/v1"
+	operatorclient "github.com/openshift/cluster-ingress-operator/pkg/operator/client"
+	"github.com/openshift/cluster-ingress-operator/pkg/operator/controller"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/rand"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"os"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+	"strings"
+	"testing"
+	"time"
+)
+
+const (
+	awsLBEIPAllocationAnnotation = "service.beta.kubernetes.io/aws-load-balancer-eip-allocations"
+)
+
+func CreateAWSEIPs() ([]string, error) {
+	kubeClient, err := createClient()
+	if err != nil {
+		return nil, fmt.Errorf("test skipped: %v", err)
+	}
+	decodedAccessKeyID, decodedSecretAccessKey, err := getIDAndKey(kubeClient)
+	if err != nil {
+		return nil, fmt.Errorf("test skipped: %v", err)
+	}
+	if infraConfig.Status.PlatformStatus == nil {
+		return nil, fmt.Errorf("test skipped on nil platform")
+	}
+	region, err := getRegion()
+	if err != nil {
+		return nil, fmt.Errorf("test skipped: %v", err)
+	}
+	clusterName, err := getClusterName()
+	if err != nil {
+		return nil, fmt.Errorf("test skipped: %v", err)
+	}
+
+	// Set up an AWS session
+	sess, err := createSession(region, decodedAccessKeyID, decodedSecretAccessKey)
+	if err != nil {
+		return nil, fmt.Errorf("test skipped: %v", err)
+	}
+
+	// Create an EC2 service client
+	svc := ec2.New(sess)
+
+	vpcID, err := getVPCID(clusterName, svc)
+	if err != nil {
+		return nil, fmt.Errorf("test skipped: %v", err)
+	}
+
+	publicSubnets, err := getPublicSubnets(vpcID, svc)
+	if err != nil {
+		return nil, fmt.Errorf("test skipped: %v", err)
+	}
+
+	// Set up random seed
+	rand.Seed(time.Now().UnixNano())
+
+	// Set tag key and value
+	tagKeyEIP := fmt.Sprintf("NE1398-%s", os.Getenv("USER"))
+	tagValueEIP := clusterName
+
+	// Allocate EIPs
+	var allocationIDs []string
+	for i := 0; i < len(publicSubnets); i++ {
+		// Allocate EIP
+		result, err := svc.AllocateAddress(&ec2.AllocateAddressInput{
```

Comment:
We should very careful about this and clean up the EIPs, I think running this in a cluster & CI is actually going to cause EIPs to get left over, and our CI bill might actually grow! (I think every EIP cost $$$).

There is a AWS reaper, but 1. I'm not sure if it works with EIPs, and 2. It may need a special tag in order to be deleted. This is probably worth looking into. Having a tag might be a good fail safe regardless.

Regardless, we should clean up. I wonder if we should even try to clean up any EIPs that are orphaned from the previous test runs. Until then, be careful about running this in your clusters, and I wouldn't run it until you have a clean up mechanism (or you know how to manually clean up. I ran this E2E and had to manually clean up the EIPs in the Openshift-dev acccount in AWS.

Reply:
done. i cleaned for the vpc with tag having cluster name.

Reply:
The logic has been revised.